README.md
# web excel

[在线预览](https://emilyyoung71415.github.io/web-excel/index.html)

<img src="./screenshots/版本1效果图.png" width="800" heght="600"/>

使用
===

直接在浏览器里引用

```html
<link href="//github.com/EmilyYoung71415/web-excel/blob/gh-pages/styles/webexcel.css" rel="stylesheet"></head>
<body onload="load()">
    <div id="app">
        <div id="webexcel"></div>
    </div>
    <script>
        function load() {
            webexcel(document.getElementById('webexcel'))
                .loadData({
                    borders: [
                        [1, 'dashed', '#0366d6'],
                    ],
                    styles: [
                        { bgcolor: 'rgb(82, 196, 26)', bi: 0, color: '#FFF' },
                    ],
                    rowm: {// 特殊列 列高
                        1: { height: 40}
                    },
                    colm: {
                        1: { width: 60 },
                        4: { width: 140 }
                    },
                    cellmm: {
                        1: {
                            1: { text: '输入长文字测试截断', si: 0 },
                        },
                    },
                })
        }
    </script>
    <script type="text/javascript" src="//github.com/EmilyYoung71415/web-excel/blob/gh-pages/webexcel.js"></script>
</body>
```

option配置

```js
// default options
{
    rowm: {// 特殊列 列高
        1: { height: 40}
    },
    colm: {
        1: { width: 60 },
        4: { width: 140 }
    },
    cellmm: {// 单元格数据
        1: {
            1: { text: '输入长文字测试截断', si: 0 },
        },
    },
    borders: [
        [1, 'dashed', '#0366d6'],
    ],
    styles: [
        { bgcolor: 'rgb(82, 196, 26)', bi: 0, color: '#FFF' },
    ],
    viewRect: { // 整个表格容器大小 = canvas大小
        height: 400,
        width: 800,
    },
    row: { // 表格初始化 10行 每行25px高
        len: 20,
        height: 25,
    },
    col: {
        len: 15,
        width: 100,
        indexWidth: 60, // 列索引栏宽度
        minWidth: 60, // 伸缩最小宽度
    },
    style: {
        bgcolor: '#ffffff',
        align: 'left',
        valign: 'top',
        wrapText: true, // 文字分行
        textDecoration: 'normal',
        color: '#333333',
        contentLineSize: 0.5,
        contentLineColor: '#d0d0d0',
        fixedheaderstyle: {
            bgcolor: '#f4f5f8',
            textAlign: 'center',
            textBaseline: 'middle',
            font: {
                size: 12,
                family: 'sans-serif',
            },
            fillStyle: '#585757',
            lineWidth: 0.5,
            strokeStyle: '#d0d0d0',
        },
        font: {
            family: 'Arial',
            size: 14,
            bold: false,
            italic: false,
        },
    },
}
```

运行
===

```bash
npm i
npm start
```

实现功能
===

版本0.1：表格基本渲染能力
- [x] 坐标网格渲染
- [x] 行列伸缩
- [x] 滚动
- [x] 单元格文字渲染处理(换行、溢出)
- [x] 选中框
- [x] 编辑框 

敬请期待：
- [ ] 合并单元格
- [ ] 复制粘贴
- [ ] 插件机制
....

原理概览
===

<img src="./screenshots/版本1架构图.png" width="800" heght="600"/>

CHANGELOG.md
遵循 [Semantic Versioning 2.0.0](http://semver.org/lang/zh-CN/) 语义化版本规范。

#### 发布周期

- 修订版本号：bugfix（当你做了不兼容的 API 修改）
- 次版本号： 新功能  （当你做了向下兼容的功能性新增）
- 主版本号：新特性   （当你做了向下兼容的问题修正）

icon说明：
🏆 里程碑
🐞 fix
🌟 亮点
💄 主题色
🚮 移除
🆕 新增
🛠 属性、api
⚡️  提升性能
🛠 重构
🔧 修改配置文件

---
## 0.6.3
`2021-04-19`
- 🌟 提升表格清晰度: Retina高清屏适配
- 🐞滚动bugfix
  - 滚动联带选中框的交互fix
  - 滚动联带索引栏hover的小方块交互fix
  - 滚动时有特殊行高时渲染行数丢失fix
- 🐞编辑bugfix
  - 编辑时选中的单元格可通过键盘键移动
  - 编辑时清除单元格为空无法清除

## 0.6.0
- 坐标网格渲染
- 行列伸缩
  -  鼠标移动到行列表头时，当前行or列高亮
  -  mousemove时，有辅助线跟随
  -  行列伸缩的附加影响：当前行列面积改变、命中的选中框面积改变
- 滚动
  - 滚动的触发：鼠标向下拖动滚动条、鼠标滚轮mousewheel
  - 表头索引栏固定不动
  - 表格内容滚动
  - 命中的单元格跟随滚动
- 选中框
  - 单击单元实现单元格选中
  - 单击单元格，并mousemove，实现单元格多选
  - 鼠标上下左右键，实现单元格移动
  - 滚动时候，命中的选中框实现跟随滚动
- 单元格文字渲染处理(换行、溢出)
  - 文字不超出单元格宽高
  - 行列伸缩时，文字要重绘，以适应单元格大小
- 编辑框
  - 双击单元格进入编辑状态，该单元格可输入文字
- viewdata
  - 所有与表格实际意义有关的数据抽象在viewdata里，改变方法也在viewdata里


src/viewdata.js
import help from './utils/help';
import onchange from './utils/onchange';
import {performanceLog} from './utils/log';
// 表格基础设置：行总数、高度；表格基本样式
const sheetOptions = {
    triggerLoaded: false,
    status: {
        peaceful: true,   // 静静地展示
        focusing: false,  // 聚焦
        editing: false,   // 编辑中： 在聚焦前提下
        scrolling: false, // 滚动中
    },
    pixelRatio: window.devicePixelRatio || 1,
    viewRect: { // 整个表格容器大小 = canvas大小
        height: 400, // window.innerHeight,
        width: 800, // window.innerWidth,
    },
    row: { // 表格初始化 10行 每行25px高
        len: 300,
        height: 25,
    },
    col: {
        len: 15,
        width: 100,
        indexWidth: 60, // 列索引栏宽度
        minWidth: 60, // 伸缩最小宽度
    },
    style: {
        bgcolor: '#ffffff',
        align: 'left',
        valign: 'top',
        wrapText: true, // 文字分行
        textDecoration: 'normal',
        color: '#333333',
        contentLineSize: 0.5,
        contentLineColor: '#d0d0d0',
        fixedheaderstyle: {
            bgcolor: '#f4f5f8',
            textAlign: 'center',
            textBaseline: 'middle',
            font: {
                size: 12,
                family: 'sans-serif',
            },
            fillStyle: '#585757',
            lineWidth: 0.5,
            strokeStyle: '#d0d0d0',
        },
        font: {
            family: 'Arial',
            size: 14,
            bold: false,
            italic: false,
        },
    },
};
// 表格状态数据：单元格数据、特殊行列setting、选区、滚动
const defaultViewData = {
    fixedColWidth: 60,
    // fixedRowHeight: 160,
    rowm: {}, // Map<int, Row>, len
    colm: {}, // Map<int, Row>, len
    cellmm: {}, // Map<int, Map<int, Cell>>
    styles: [], // RangeMan
    borders: [], // // RangeMan
    scrollOffset: {x: 0, y: 0}, // 滚动物理距离
    scrollIndexes: [0, 0], // [ri, ci] rowi, coli 由滚动的物理距离计算出 滚动了多少单元格个数（行个数、列个数）
    selectRectIndexes: null, // [[sri, sci],[eri, eci]] startRowI,
    // render: () => { }, // viewData改变之后 视图重新渲染
    // addHistory: () => { },
};

function proxyData(data) {
    Object.keys(data).forEach(key => {
        Object.defineProperty(this, key, {
            get() {
                return data[key];
            },
            set(newValue) {
                data[key] = newValue;
            },
        });
    });
}

function changeHandler(path, value, prevValue) {
    if (!this.triggerLoaded) {
        return;
    }
    // this 完整的viewdata对象
    // 将历史记录放入 dep 将table的render放入
    console.log('%c viewdatachanged', 'color:blue', {
        path,
        value: JSON.stringify(value),
        prevValue: JSON.stringify(prevValue),
    });
    performanceLog();
}
export default class ViewData {
    // sheet defaultoptions
    constructor(options) {
        const data = Object.assign(sheetOptions, defaultViewData, options);
        this.data = onchange(data, changeHandler);
        // viewdata的访问代理
        proxyData.call(this, this.data);// this.data.row 都可以通过 this.row访问
    }

    // 在constructor 和 loaddata里 都分别调用了render()
    loadData(data) {
        this.data = Object.assign(this.data, data, {
            triggerLoaded: true,
        });
    }

    colTotalWidth() {
        const {col, colm} = this;
        const [cmTotal, cmSize] = help.sum(colm, v => v.width || 0);
        return ((col.len - cmSize) * col.width) + cmTotal;
    }

    rowTotalHeight() {
        const {row, rowm} = this;
        const [rmTotal, rmSize] = help.sum(rowm, v => v.height || 0);
        return ((row.len - rmSize) * row.height) + rmTotal;
    }

    /**
     * 从表头开始遍历，到第rowlen行
     * 每经过一行，返回当前行：行index、行offsetY,当前行的行高
     * @param {*} rowLen 第rowLen行
     * @param {*} cb(index, offsetY, curRowHeight)
     */
    rowEach(rowLen, cb) {
        let y = 0;
        for (let i = 0; i <= rowLen; i += 1) {
            const rowHeight = this.getRowHeight(i);
            cb && cb(i, y, rowHeight);
            y += rowHeight;
        }
        return y;
    }

    colEach(colLen, cb) {
        let x = 0;
        for (let i = 0; i <= colLen; i += 1) {
            const colWidth = this.getColWidth(i);// 当前列的列宽
            // 列索引 当前一笔的起始点 当前列的列宽
            cb && cb(i, x, colWidth);
            x += colWidth;
        }
        return x;
    }

    getColWidth(index) { // 如果当前列未在特殊样式列 就是normal-size
        const {col, colm} = this;
        return colm[`${index}`] ? colm[`${index}`].width : col.width;
    }

    getRowHeight(index) {
        const {row, rowm} = this;
        return rowm[`${index}`] ? rowm[`${index}`].height : row.height;
    }

    // 计算选中的这一坨列的宽度
    colSumWidth(starIndex, endIndex) {
    // 传入table的索引 min max,获取每列宽度的函数
    // 返回当前范围 [index_colmin,index_max] 的列宽
        return help.rangeSum(starIndex, endIndex, i => this.getColWidth(i));
    }

    rowSumHeight(starIndex, endIndex) {
        return help.rangeSum(starIndex, endIndex, i => this.getRowHeight(i));
    }

    /**
     * 根据当前selector的 逻辑索引 [start单元格(行index，列index)， end单元格（行index，列index）]
     * 得到selector逻辑索引对应的 物理left\top
     * @returns
     */
    getSelectRect() {
        const {scrollOffset} = this;
        // FIX: 初始化未产生选中框时 selector会报undefined is not iterable
        if (!this.selectRectIndexes) {
            return;
        }
        const [[sri, sci], [eri, eci]] = this.selectRectIndexes || [];
        const {left, top} = this.cellPosition(sri - 1, sci - 1);
        let height = this.rowSumHeight(sri - 1, eri);
        let width = this.colSumWidth(sci - 1, eci);

        if (eri >= 0 && eci === 0) {
            width = this.colTotalWidth();
        }
        if (eri === 0 && eci >= 0) {
            height = this.rowTotalHeight();
        }
        return {
            left: left - scrollOffset.x,
            top: top - scrollOffset.y,
            height,
            width,
        };
    }

    /**
     * 单元格逻辑索引 ->  物理定位 left、top
     * @param {*} ri 行index
     * @param {*} ci 列index
     * @returns 当前单元格物理定位
     */
    cellPosition(ri, ci) {
        const left = this.colSumWidth(0, ci);
        const top = this.rowSumHeight(0, ri);
        return {
            left, top,
        };
    }

    /**
     * 维护变量：最新第一行索引=ri+1 scrollIndexes [ri, ci]
     * @param {*} offset 滚动的距离
     */
    scroll(offset) {
        const {x, y} = offset;// 滚动条在浏览器中 横向滚动距离 和 竖向滚动距离
        const {scrollOffset, col, row} = this;
        // y:y>0 都要产生滚动距离 按照滚动区间来 划分真正的滚动距离
        if (y || y === 0) {
            // 由实际滚动距离算出，应该滚动的距离（帮用户按照一行一行地滚）
            // 每次滚动是一格一格的滚动 不存在中间差
            // top是不加上当前rowheight的滚动距离 实际上 top+height是真正的滚动距离
            const [ri, top, curRowHeight] = help.rangeReduceIf(
                0,
                row.len,
                0,
                0,
                y, // 要达到y: scrollTop的滚动距离 需要滚动多少单元行
                i => this.getRowHeight(i)
            );
            let y1 = top;
            if (y > 0) {
                y1 += curRowHeight;
            }
            // 只要产生了滚动都至少会滚动一行
            this.scrollIndexes[0] = y > 0 ? ri : 0;// ri+1是最新第一行索引
            scrollOffset.y = y1 * this.pixelRatio;
        }
        if (x || x === 0) {
            const [ci, left, curColWidth] = help.rangeReduceIf(
                0,
                col.len,
                0,
                0,
                x,
                i => this.getColWidth(i)
            );
            let x1 = left;
            if (x > 0) {
                x1 += curColWidth;
            }
            this.scrollIndexes[1] = x > 0 ? ci : 0;
            scrollOffset.x = x1 * this.pixelRatio;
        }
    }

    // 根据鼠标坐标点，获得所在的cell矩形信息
    // (ri, ci, offsetX, offsetY, width, height)
    getCellRectWithIndexes(x, y, forSelector = true) {
        const {ri, top, height} = this.getCellRowByY(y);
        const {ci, left, width} = this.getCellColByX(x);
        const {row, col} = this;
        // 鼠标坐标在 第一列 索引栏
        // 如果是forSelector就是行选 返回的width是table_width
        if (ri >= 0 && ci === 0) {
            const nwidth = forSelector ? this.colTotalWidth() : width;
            const ntop = forSelector ? top - row.height : top;
            return {
                ri, ci, left: 0, top: ntop, width: nwidth, height,
            };
        }
        if (ri === 0 && ci >= 0) {
            const nheight = forSelector ? this.rowTotalHeight() : height;
            const nleft = forSelector ? left - col.indexWidth : left;
            return {
                ri, ci, left: nleft, top: 0, width, height: nheight,
            };
        }
        return {
            ri, ci, left: left - col.indexWidth, top: top - row.height, width, height,
        };
    }

    getCellRowByY(y) { // 根据y坐标获得所在行号
        const {row, scrollOffset} = this;
        const [ri, top, height] = help.rangeReduceIf(
            0,
            row.len,
            row.height - scrollOffset.y, // top
            row.height, // 行高
            y,
            i => this.getRowHeight(i) // 传入获取第i行行高的cb函数
        );
        if (top <= 0) {
            return {ri: 0, top: 0, height};
        }
        return {ri, top, height};
    }

    getCellColByX(x) {
        const {col, scrollOffset} = this;
        const [ci, left, width] = help.rangeReduceIf(
            0,
            col.len,
            col.indexWidth - scrollOffset.x,
            col.indexWidth,
            x,
            i => this.getColWidth(i)
        );
        if (left <= 0) {
            return {ci: 0, left: 0, width: col.indexWidth};
        }
        return {ci, left, width};
    }

    setRowHeight(index, height) {
        this.rowm[index] = this.rowm[index] || {};
        this.rowm[index].height = height;
    }

    setColWidth(index, width) {
        this.colm[index] = this.colm[index] || {};
        this.colm[index].width = width;
    }

    setSelectRectIndexes(index) {
        this.selectRectIndexes = index;
        return this;
    }

    getCell(ri, ci) {
        if (this.cellmm[ri] && this.cellmm[ri][ci]) {
            return this.cellmm[ri][ci];
        }
        return null;
    }

    // 编辑完成 input的文字放进canvas里
    setCellText(ri, ci, itext) {
        const {cellmm} = this;
        cellmm[ri] = cellmm[ri] || {};
        cellmm[ri][ci] = cellmm[ri][ci] || {};
        cellmm[ri][ci].text = itext;
    }
}
